File: loadercomplex-forge-18/src/main/java/com/enderzombi102/loadercomplex/forge18/impl/block/ForgeBlock.java
package com.enderzombi102.loadercomplex.forge18.impl.block;

import com.enderzombi102.loadercomplex.api.block.Block;
import com.enderzombi102.loadercomplex.api.utils.Callable;
import com.enderzombi102.loadercomplex.forge18.impl.entity.ForgeEntity;
import com.enderzombi102.loadercomplex.forge18.impl.entity.ForgePlayer;
import com.enderzombi102.loadercomplex.forge18.impl.utils.BlockUtils;
import com.enderzombi102.loadercomplex.forge18.impl.world.ForgeWorld;
import net.minecraft.block.BlockState;
import net.minecraft.block.Material;
import net.minecraft.entity.Entity;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.server.world.ServerWorld;
import net.minecraft.sound.BlockSoundGroup;
import net.minecraft.util.ActionResult;
import net.minecraft.util.Hand;
import net.minecraft.util.hit.BlockHitResult;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.BlockView;
import net.minecraft.world.World;
import org.jetbrains.annotations.NotNull;

import java.util.Random;

public class ForgeBlock extends net.minecraft.block.Block {

	private final Block blockImpl;

	public ForgeBlock(Block block) {
		super(
			( (Callable<Settings>) () -> {
				var settings =  Settings.of( Material.STONE )
					.slipperiness( block.slipperiness )
					.resistance( block.resistance )
					.strength( block.hardness )
					.luminance( state -> block.lightLevel );

				if (! block.hasCollision )
					settings.noCollision();

				if (! block.opaque )
					settings.nonOpaque();

				if ( block.randomTicks )
					settings.ticksRandomly();

				return settings;
			} ).call()
		);
		this.blockImpl = block;
		block.implementationBlock = this;
	}

	// logic method overrides


	public boolean isEqualTo(net.minecraft.block.Block block) {
		return block instanceof ForgeBlock && ( (ForgeBlock) block ).blockImpl == this.blockImpl;
	}

	@Override
	public @NotNull ActionResult onUse(@NotNull BlockState state, @NotNull World world, @NotNull BlockPos pos, @NotNull PlayerEntity player, Hand hand, BlockHitResult hit) {
		return this.blockImpl.OnBlockInteracted(
			new ForgeWorld( world ),
			BlockUtils.toPosition( pos ),
			new ForgeBlockstate( state ),
			new ForgePlayer( player ),
			com.enderzombi102.loadercomplex.api.utils.Hand.valueOf( hand.name() ),
			com.enderzombi102.loadercomplex.api.utils.Direction.valueOf( hit.getSide().name() ),
			hit.getBlockPos().getX(), hit.getBlockPos().getY(), hit.getBlockPos().getZ()
		) ? ActionResult.SUCCESS : ActionResult.PASS;
	}

	@Override
	public void onSteppedOn(@NotNull World world, @NotNull BlockPos pos, @NotNull BlockState state, @NotNull Entity entity) {
		this.blockImpl.OnSteppedOn(
			new ForgeWorld( world ),
			BlockUtils.toPosition( pos ),
			new ForgeEntity( entity )
		);
	}

	@Override
	public void onBreak(World world, BlockPos pos, BlockState state, PlayerEntity player) {
		this.blockImpl.OnBreak(
			new ForgeWorld( world ),
			BlockUtils.toPosition( pos ),
			new ForgeBlockstate( state ),
			new ForgePlayer( player )
		);
	}

	@Override
	public void onEntityCollision(BlockState state, World world, BlockPos pos, Entity entity) {
		this.blockImpl.OnEntityCollision(
			new ForgeWorld( world ),
			BlockUtils.toPosition( pos ),
			new ForgeBlockstate( state ),
			new ForgeEntity( entity )
		);
	}

	@Override
	public void randomTick(BlockState state, ServerWorld world, BlockPos pos, Random random) {
		this.blockImpl.OnRandomTick(
			new ForgeWorld( world ),
			BlockUtils.toPosition( pos ),
			new ForgeBlockstate( state ),
			random
		);
	}

	// getter methods overrides

	@Override
	public BlockSoundGroup getSoundGroup(BlockState state) {
		return switch (this.blockImpl.soundGroup) {
			case WOOD -> BlockSoundGroup.WOOD;
			case GRAVEL -> BlockSoundGroup.GRAVEL;
			case GRASS -> BlockSoundGroup.GRASS;
			default -> /* case STONE: */ BlockSoundGroup.STONE;
			case METAL -> BlockSoundGroup.METAL;
			case GLASS -> BlockSoundGroup.GLASS;
			case CLOTH -> BlockSoundGroup.WOOL;
			case SAND -> BlockSoundGroup.SAND;
			case SNOW -> BlockSoundGroup.SNOW;
			case LADDER -> BlockSoundGroup.LADDER;
			case ANVIL -> BlockSoundGroup.ANVIL;
			case SLIME -> BlockSoundGroup.SLIME;
		};
	}

	@Override
	public int getOpacity(BlockState state, BlockView world, BlockPos pos) {
		return this.blockImpl.opacity;
	}

	@Override
	public boolean isTranslucent(BlockState state, BlockView world, BlockPos pos) {
		return this.blockImpl.translucent;
	}

	@Override
	public float getBlastResistance() {
		return this.blockImpl.resistance / 5.0F;
	}
}

File: loadercomplex-forge/src/ver182/kotlin/com/enderzombi102/loadercomplex/forge182/impl/block/ForgeBlock.kt
package com.enderzombi102.loadercomplex.forge182.impl.block

import com.enderzombi102.loadercomplex.api.math.Direction
import com.enderzombi102.loadercomplex.api.minecraft.block.Block
import com.enderzombi102.loadercomplex.api.minecraft.block.BlockSoundGroup
import com.enderzombi102.loadercomplex.api.minecraft.util.Hand
import com.enderzombi102.loadercomplex.forge182.impl.utils.toLC
import com.enderzombi102.loadercomplex.forge182.impl.entity.ForgeEntity
import com.enderzombi102.loadercomplex.forge182.impl.entity.ForgePlayer
import com.enderzombi102.loadercomplex.forge182.impl.utils.toMC
import com.enderzombi102.loadercomplex.forge182.impl.world.ForgeWorld
import net.minecraft.block.BlockState
import net.minecraft.block.Material
import net.minecraft.entity.Entity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.server.world.ServerWorld
import net.minecraft.util.ActionResult
import net.minecraft.util.hit.BlockHitResult
import net.minecraft.util.math.BlockPos
import net.minecraft.world.BlockView
import net.minecraft.world.World
import java.util.*
import net.minecraft.sound.BlockSoundGroup as McBlockSoundGroup
import net.minecraft.util.Hand as McHand


@Suppress("OVERRIDE_DEPRECATION")
class ForgeBlock(private val blockImpl: Block) : net.minecraft.block.Block( createSettings(blockImpl) ) {
	init {
		blockImpl.implementationBlock = this
	}

	// logic method overrides
	fun isEqualTo(block: net.minecraft.block.Block?): Boolean {
		return block is ForgeBlock && block.blockImpl === this.blockImpl
	}

	override fun onUse(state: BlockState, world: World, pos: BlockPos, player: PlayerEntity, hand: McHand, hit: BlockHitResult ): ActionResult {
		return blockImpl.onBlockInteracted(
			ForgeWorld(world),
			pos.toLC(),
			ForgeBlockState(state),
			ForgePlayer(player),
			Hand.valueOf(hand.name),
			Direction.valueOf(hit.side.name),
			hit.pos.x, hit.pos.y, hit.pos.z
		).toMC()
	}

	override fun onSteppedOn(world: World, pos: BlockPos, state: BlockState, entity: Entity) {
		blockImpl.onSteppedOn(
			ForgeWorld(world),
			pos.toLC(),
			ForgeEntity(entity)
		)
	}

	override fun onBreak(world: World, pos: BlockPos, state: BlockState, player: PlayerEntity) {
		blockImpl.onBreak(
			ForgeWorld(world),
			pos.toLC(),
			ForgeBlockState(state),
			ForgePlayer(player)
		)
	}

	override fun onEntityCollision(state: BlockState, world: World, pos: BlockPos, entity: Entity) {
		blockImpl.onEntityCollision(
			ForgeWorld(world),
			pos.toLC(),
			ForgeBlockState(state),
			ForgeEntity(entity)
		)
	}

	override fun randomTick(state: BlockState, world: ServerWorld, pos: BlockPos, random: Random) {
		blockImpl.onRandomTick(
			ForgeWorld(world),
			pos.toLC(),
			ForgeBlockState(state),
			random
		)
	}

	// getter methods overrides
	override fun getSoundGroup(state: BlockState): McBlockSoundGroup {
		return when (blockImpl.soundGroup) {
			BlockSoundGroup.WOOD -> McBlockSoundGroup.WOOD
			BlockSoundGroup.GRAVEL -> McBlockSoundGroup.GRAVEL
			BlockSoundGroup.GRASS -> McBlockSoundGroup.GRASS
			BlockSoundGroup.METAL -> McBlockSoundGroup.METAL
			BlockSoundGroup.GLASS -> McBlockSoundGroup.GLASS
			BlockSoundGroup.CLOTH -> McBlockSoundGroup.WOOL
			BlockSoundGroup.SAND -> McBlockSoundGroup.SAND
			BlockSoundGroup.SNOW -> McBlockSoundGroup.SNOW
			BlockSoundGroup.LADDER -> McBlockSoundGroup.LADDER
			BlockSoundGroup.ANVIL -> McBlockSoundGroup.ANVIL
			BlockSoundGroup.SLIME -> McBlockSoundGroup.SLIME
			else      /* STONE */ -> McBlockSoundGroup.STONE
		}
	}

	override fun getOpacity(state: BlockState, world: BlockView, pos: BlockPos): Int {
		return blockImpl.opacity
	}

	override fun isTranslucent(state: BlockState, world: BlockView, pos: BlockPos): Boolean {
		return blockImpl.translucent
	}

	override fun getBlastResistance(): Float {
		return blockImpl.resistance / 5.0f
	}

	companion object {
		private fun createSettings( blockImpl: Block ): Settings {
			val settings = Settings.of( Material.STONE )
				.slipperiness( blockImpl.slipperiness )
				.resistance( blockImpl.resistance )
				.strength( blockImpl.hardness )
				.luminance { blockImpl.lightLevel }
			if (! blockImpl.hasCollision ) {
				settings.noCollision()
			}

			if (! blockImpl.opaque ) {
				settings.nonOpaque()
			}

			if ( blockImpl.randomTicks ) {
				settings.ticksRandomly()
			}
			return settings
		}
	}
}
